File: agentmail-note-taker/typescript/src/notesStore.ts
/**
 * Markdown-with-frontmatter notes store.
 *
 * Each note is a single Markdown file in `notes/<YYYY-MM-DD>-<slug>.md` with
 * YAML frontmatter. The file is the source of truth — re-read on every search.
 */

import { existsSync, readFileSync, writeFileSync, mkdirSync, readdirSync } from "node:fs";
import { join } from "node:path";

const NOTES_DIR = "notes";

function slugify(text: string, maxLen = 40): string {
  return text.toLowerCase().trim()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-+|-+$/g, "")
    .slice(0, maxLen) || "note";
}

function yamlFrontmatter(meta: Record<string, any>): string {
  const lines = ["---"];
  for (const [k, v] of Object.entries(meta)) {
    if (Array.isArray(v)) {
      lines.push(`${k}: [${v.join(", ")}]`);
    } else {
      const sv = String(v).replace(/"/g, '\\"');
      if (/[:,#\[\]]/.test(sv)) lines.push(`${k}: "${sv}"`);
      else lines.push(`${k}: ${sv}`);
    }
  }
  lines.push("---");
  return lines.join("\n");
}

export function findByThread(threadId: string): string | null {
  if (!existsSync(NOTES_DIR) || !threadId) return null;
  for (const fn of readdirSync(NOTES_DIR)) {
    if (!fn.endsWith(".md")) continue;
    const path = join(NOTES_DIR, fn);
    try {
      const head = readFileSync(path, "utf-8").slice(0, 2000);
      if (head.includes(`thread_id: ${threadId}`) || head.includes(`thread_id: "${threadId}"`)) {
        return path;
      }
    } catch { /* skip */ }
  }
  return null;
}

export interface NoteInput {
  sourceSummary: string;
  threadId: string;
  tags: string[];
  summary: string;
  decisions: string[];
  actionItems: Array<{ owner?: string; task?: string; deadline?: string; urgency?: string }>;
  openQuestions: string[];
  keyFacts: string[];
  existingPath?: string | null;
}

export function writeNote(input: NoteInput): string {
  if (!existsSync(NOTES_DIR)) mkdirSync(NOTES_DIR, { recursive: true });
  const today = new Date().toISOString().slice(0, 10);

  let path: string;
  if (input.existingPath && existsSync(input.existingPath)) {
    path = input.existingPath;
  } else {
    const slugSeed = ((input.tags[0] || "") + " " + input.summary.slice(0, 40));
    let candidate = join(NOTES_DIR, `${today}-${slugify(slugSeed)}.md`);
    let i = 1;
    while (existsSync(candidate)) {
      candidate = join(NOTES_DIR, `${today}-${slugify(slugSeed)}-${i}.md`);
      i++;
    }
    path = candidate;
  }

  const meta = {
    source: input.sourceSummary,
    date: today,
    thread_id: input.threadId || "",
    tags: input.tags,
  };

  const lines: string[] = [yamlFrontmatter(meta), "", `# Summary`, "", input.summary, ""];

  if (input.decisions.length) {
    lines.push("## Decisions", "");
    for (const d of input.decisions) lines.push(`- ${d}`);
    lines.push("");
  }
  if (input.actionItems.length) {
    lines.push("## Action items", "");
    for (const ai of input.actionItems) {
      const owner = ai.owner || "(unassigned)";
      const task = ai.task || "";
      const tail = [ai.deadline, ai.urgency].filter(Boolean).join(" · ");
      const tailStr = tail ? `  (${tail})` : "";
      lines.push(`- **${owner}**: ${task}${tailStr}`);
    }
    lines.push("");
  }
  if (input.openQuestions.length) {
    lines.push("## Open questions", "");
    for (const q of input.openQuestions) lines.push(`- ${q}`);
    lines.push("");
  }
  if (input.keyFacts.length) {
    lines.push("## Key facts", "");
    for (const f of input.keyFacts) lines.push(`- ${f}`);
    lines.push("");
  }

  writeFileSync(path, lines.join("\n"), "utf-8");
  return path;
}

export interface NoteMeta {
  path: string;
  date: string;
  tags: string[];
  threadId: string;
  source: string;
  summary: string;
}

export function parseFrontmatter(text: string): Record<string, any> {
  if (!text.startsWith("---")) return {};
  const end = text.indexOf("\n---", 3);
  if (end === -1) return {};
  const block = text.slice(3, end);
  const out: Record<string, any> = {};
  for (const line of block.trim().split("\n")) {
    if (!line.includes(":")) continue;
    const [k, ...rest] = line.split(":");
    const v = rest.join(":").trim();
    if (v.startsWith("[") && v.endsWith("]")) {
      out[k.trim()] = v.slice(1, -1).split(",").map(x => x.trim().replace(/^"|"$/g, "")).filter(Boolean);
    } else if (v.startsWith('"') && v.endsWith('"')) {
      out[k.trim()] = v.slice(1, -1);
    } else {
      out[k.trim()] = v;
    }
  }
  return out;
}

export function listAll(): NoteMeta[] {
  if (!existsSync(NOTES_DIR)) return [];
  const files = readdirSync(NOTES_DIR).filter(f => f.endsWith(".md")).sort().reverse();
  const items: NoteMeta[] = [];
  for (const fn of files) {
    const path = join(NOTES_DIR, fn);
    let text: string;
    try { text = readFileSync(path, "utf-8"); } catch { continue; }
    const meta = parseFrontmatter(text);
    let summary = "";
    const afterFm = text.split("---", 3).slice(2).join("---") || text;
    const m = afterFm.match(/# Summary\n+([\s\S]+?)(?:\n\n|\n#)/);
    if (m) summary = m[1].trim().replace(/\n/g, " ");
    items.push({
      path, date: meta.date || "", tags: meta.tags || [],
      threadId: meta.thread_id || "", source: meta.source || "", summary,
    });
  }
  return items;
}

export function search(query: string, limit = 10): NoteMeta[] {
  const items = listAll();
  if (!items.length) return [];
  const qTerms = query.toLowerCase().split(/\W+/).filter(t => t.length > 2);
  const scored: Array<[number, NoteMeta]> = [];
  for (const item of items) {
    const text = (item.summary + " " + item.tags.join(" ") + " " + item.source).toLowerCase();
    let body = "";
    try { body = readFileSync(item.path, "utf-8").toLowerCase(); } catch {}
    let score = 0;
    for (const t of qTerms) {
      score += (text.match(new RegExp(t, "g")) || []).length * 3
             + (body.match(new RegExp(t, "g")) || []).length;
      if (item.tags.includes(t)) score += 5;
    }
    if (score > 0) scored.push([score, item]);
  }
  scored.sort((a, b) => b[0] - a[0]);
  return scored.slice(0, limit).map(([_, item]) => item);
}

export function readNoteExcerpt(path: string, maxChars = 1500): string {
  if (!existsSync(path)) return "";
  let text = readFileSync(path, "utf-8");
  if (text.startsWith("---")) {
    const end = text.indexOf("\n---", 3);
    if (end !== -1) text = text.slice(end + 4).replace(/^\s+/, "");
  }
  return text.slice(0, maxChars);
}

File: python/notes_store.py
"""
Markdown-with-frontmatter notes store.

Each note is a single Markdown file in `notes/<YYYY-MM-DD>-<slug>.md`. The
file is the source of truth — the agent always re-reads from disk so the
user can edit notes by hand and the agent will pick up the changes.

Frontmatter (YAML) at the top of each file:
    ---
    source: "Fwd from Sarah Chen, 2026-04-29"
    date: "2026-04-29"
    thread_id: "<message-thread-id-from-agentmail>"
    tags: [work, q3-planning]
    ---

Body is the structured content (DECISIONS / ACTION ITEMS / OPEN QUESTIONS /
KEY FACTS), rendered for human readability.
"""

import re
from datetime import datetime
from pathlib import Path

NOTES_DIR = Path("notes")


def _slugify(text: str, max_len: int = 40) -> str:
    text = text.lower().strip()
    text = re.sub(r"[^a-z0-9]+", "-", text)
    text = text.strip("-")
    return text[:max_len] or "note"


def _yaml_frontmatter(meta: dict) -> str:
    lines = ["---"]
    for k, v in meta.items():
        if isinstance(v, list):
            lines.append(f"{k}: [{', '.join(str(x) for x in v)}]")
        else:
            # Quote string values that contain special chars
            sv = str(v).replace('"', '\\"')
            if any(c in sv for c in ":,#[]"):
                lines.append(f'{k}: "{sv}"')
            else:
                lines.append(f"{k}: {sv}")
    lines.append("---")
    return "\n".join(lines)


def find_by_thread(thread_id: str) -> Path | None:
    """Return the existing note for this thread, if any (for dedup)."""
    if not NOTES_DIR.exists() or not thread_id:
        return None
    for p in NOTES_DIR.glob("*.md"):
        try:
            head = p.read_text(encoding="utf-8")[:2000]
        except Exception:
            continue
        if f'thread_id: {thread_id}' in head or f'thread_id: "{thread_id}"' in head:
            return p
    return None


def write_note(*, source_summary: str, thread_id: str, tags: list[str],
               summary: str, decisions: list[str], action_items: list[dict],
               open_questions: list[str], key_facts: list[str],
               existing_path: Path | None = None) -> Path:
    """Write a note file. If existing_path is given, overwrites it (dedup).
    Returns the path."""
    NOTES_DIR.mkdir(exist_ok=True)
    today = datetime.now().strftime("%Y-%m-%d")

    if existing_path and existing_path.exists():
        path = existing_path
    else:
        slug_seed = (tags[0] if tags else "") + " " + summary[:40]
        path = NOTES_DIR / f"{today}-{_slugify(slug_seed)}.md"
        i = 1
        while path.exists():
            path = NOTES_DIR / f"{today}-{_slugify(slug_seed)}-{i}.md"
            i += 1

    meta = {
        "source": source_summary,
        "date": today,
        "thread_id": thread_id or "",
        "tags": tags,
    }

    body_lines = [_yaml_frontmatter(meta), "", f"# Summary", "", summary, ""]

    if decisions:
        body_lines += ["## Decisions", ""] + [f"- {d}" for d in decisions] + [""]

    if action_items:
        body_lines += ["## Action items", ""]
        for ai in action_items:
            owner = ai.get("owner") or "(unassigned)"
            task = ai.get("task") or ""
            deadline = ai.get("deadline") or ""
            urgency = ai.get("urgency") or ""
            tail = " · ".join(x for x in (deadline, urgency) if x)
            tail_str = f"  ({tail})" if tail else ""
            body_lines.append(f"- **{owner}**: {task}{tail_str}")
        body_lines.append("")

    if open_questions:
        body_lines += ["## Open questions", ""] + [f"- {q}" for q in open_questions] + [""]

    if key_facts:
        body_lines += ["## Key facts", ""] + [f"- {f}" for f in key_facts] + [""]

    path.write_text("\n".join(body_lines), encoding="utf-8")
    return path


def parse_frontmatter(text: str) -> dict:
    """Cheap YAML-ish parser for our own frontmatter (we control the format)."""
    if not text.startswith("---"):
        return {}
    end = text.find("\n---", 3)
    if end == -1:
        return {}
    block = text[3:end]
    out: dict = {}
    for line in block.strip().splitlines():
        if ":" not in line:
            continue
        k, v = line.split(":", 1)
        k = k.strip()
        v = v.strip()
        if v.startswith("[") and v.endswith("]"):
            out[k] = [x.strip().strip('"') for x in v[1:-1].split(",") if x.strip()]
        elif v.startswith('"') and v.endswith('"'):
            out[k] = v[1:-1]
        else:
            out[k] = v
    return out


def list_all() -> list[dict]:
    """Return all notes' metadata + summary line, sorted newest first."""
    if not NOTES_DIR.exists():
        return []
    items = []
    for p in sorted(NOTES_DIR.glob("*.md"), reverse=True):
        try:
            text = p.read_text(encoding="utf-8")
        except Exception:
            continue
        meta = parse_frontmatter(text)
        # First non-empty line after the body's "# Summary" header
        summary = ""
        after_fm = text.split("---", 2)[-1] if text.count("---") >= 2 else text
        m = re.search(r"# Summary\n+(.+?)(?:\n\n|\n#)", after_fm, re.DOTALL)
        if m:
            summary = m.group(1).strip().replace("\n", " ")
        items.append({
            "path": str(p),
            "date": meta.get("date", ""),
            "tags": meta.get("tags", []),
            "thread_id": meta.get("thread_id", ""),
            "source": meta.get("source", ""),
            "summary": summary,
        })
    return items


def search(query: str, limit: int = 10) -> list[dict]:
    """Cheap keyword + tag search. For >100 notes, swap for embeddings."""
    items = list_all()
    if not items:
        return []
    q_terms = [t for t in re.split(r"\W+", query.lower()) if len(t) > 2]
    scored = []
    for item in items:
        text = (item["summary"] + " " + " ".join(item["tags"]) + " " + item["source"]).lower()
        # Also pull the body for full-text
        try:
            body = Path(item["path"]).read_text(encoding="utf-8").lower()
        except Exception:
            body = ""
        score = 0
        for t in q_terms:
            score += text.count(t) * 3 + body.count(t)
            if t in item["tags"]:
                score += 5
        if score > 0:
            scored.append((score, item))
    scored.sort(reverse=True, key=lambda x: x[0])
    return [item for _, item in scored[:limit]]


def read_note_excerpt(path: str, max_chars: int = 1500) -> str:
    """Read a note's body (after frontmatter) for inclusion in a Claude turn."""
    p = Path(path)
    if not p.exists():
        return ""
    text = p.read_text(encoding="utf-8")
    if text.startswith("---"):
        end = text.find("\n---", 3)
        if end != -1:
            text = text[end + 4:].lstrip()
    return text[:max_chars]
